tests/integration_tests/reaper.py

```diff
@@ -0,0 +1,198 @@
+            )
+            return False
+
+    def _reaper_loop(self) -> None:
+        """reaper_loop() manages all instances that have been reaped
+
+        tasks:
+        - destroy newly reaped instances
+        - manage a ledger undead instances
+        - periodically attempt to kill undead instances
+        - die when instructed to
+        - ensure that every reaped instance is destroyed at least once before
+          reaper dies
+        """
+        LOG.info("Reaper: exalted in life, to assist others in death")
+        while True:
+            # nap until woken or timeout
+            with self.wake_reaper:
+                self.wake_reaper.wait(timeout=self.timeout)
+            if self._do_reap():
+                break
+        LOG.info("Reaper: exited")
+
+    def _do_reap(self) -> bool:
+        """_do_reap does a single pass of the reaper loop
+
+        return True if the loop should exit
+        """
+
+        new_undead_instances: List[IntegrationInstance] = []
+
+        # first destroy all newly reaped instances
+        while not self.reaped_instances.empty():
+            instance = self.reaped_instances.get_nowait()
+            success = self._destroy(instance)
+            if not success:
+                LOG.warning(
+                    "Reaper: failed to destroy %s",
+                    instance.instance.id,
+                )
+                # failure to delete, add to the ledger
+                new_undead_instances.append(instance)
+            else:
+                LOG.info("Reaper: destroyed %s", instance.instance.id)
+
+        # every instance has tried at least once and the reaper has been
+        # instructed to tear down - so do it
+        if self.exit_reaper.is_set():
+            if not self.reaped_instances.empty():
+                # race: an instance was added to the queue after iteration
+                # completed. Destroy the latest instance.
+                self._update_undead_ledger(new_undead_instances)
+                return False
+            self._update_undead_ledger(new_undead_instances)
+            LOG.info("Reaper: exiting")
+            if self.undead_ledger:
+                # undead instances exist - unclean teardown
+                LOG.info(
+                    "Reaper: the faults of incompetent abilities will be "
+                    "consigned to oblivion, as myself must soon be to the "
+                    "mansions of rest."
+                )
+                warnings.warn(f"Test instance(s) leaked: {self.undead_ledger}")
+            else:
+                LOG.info("Reaper: duties complete, my turn to rest")
+            LOG.info(
+                "Reaper: reaped %s/%s instances",
+                self.counter,
+                self.counter + len(self.undead_ledger),
+            )
+            return True
+
+        # attempt to destroy all instances which previously refused to
+        # destroy
+        for instance in self.undead_ledger:
```

Comment:
This seems like an unlikely situation, but if we happen to get a buildup of undead instances, then we could hit a scenario where we finished testing right after we started this list and will sit waiting for a bunch of instances to die that will likely never die.

Think it's worth a `reaped_instances.empty()` check for every new item in the for loop?

Reply:
> scenario where we finished testing right after we started this list and will sit waiting for a bunch of instances to die that will likely never die.

Yeah, that could happen. Good point.

> Think it's worth a reaped_instances.empty() check for every new item in the for loop?

That alone wouldn't work, I suspect. A majority of the time when this code is running reaped_instances.empty() is probably true, since we removed all instanced from this queue just before running this code. But if `self.exit_reaper.is_set() and self.reaped_instances.empty()` then we can probably exit early to avoid the extra waiting.